var/spack/repos/builtin/packages/hypre/package.py

```diff
@@ -48,67 +51,224 @@ class Hypre(Package):
+            common_flags = list(cuda_arch_flags)
+            if '+shared' in spec:
+                common_flags += ['-Xcompiler', self.compiler.pic_flag]
+            if '+openmp' in spec:
+                common_flags += ['-Xcompiler', self.compiler.openmp_flag]
+            cflags += common_flags
+            cuflags += common_flags
+
+            cflags_joined = ' '.join(cflags + cflags_orig)
+            cxxflags_joined = ' '.join(cflags + cxxflags)
+            make_args += [
+                'CC=%s' % nvcc,
+                'CXX=%s' % nvcc,
+                'NVCC=%s' % nvcc,
+                'CFLAGS=%s $(C_COMPILE_FLAGS)' % cflags_joined,
+                'CXXFLAGS=%s $(CXX_COMPILE_FLAGS)' % cxxflags_joined,
+                'NVCCFLAGS=%s' % ' '.join(cuflags)]
+
+            ldflags += [self.hypre_cuda_libs.ld_flags]
+
+            if '+shared' in spec:
+                # Additional options for linking a shared library
+                # ${BUILD_CC_SHARED} -o ${SONAME} ${FILES_HYPRE} ${SOLIBS}
+                #   ${SHARED_SET_SONAME}${SONAME} ${SHARED_OPTIONS} ${LDFLAGS}
+                make_args += [
+                    'BUILD_CC_SHARED=%s -shared -ccbin=%s' %
+                    (' '.join([nvcc] + cuda_arch_flags), env['CXX']),
+                    'SHARED_SET_SONAME=-Xlinker -soname,',
+                    'SHARED_OPTIONS=',
+                    'LIBS=']
+                # Patch src/struct_ls/Makefile to link with
+                # ../utilities/random.o
+                with working_dir(join_path('src', 'struct_ls')):
+                    filter_file('^(OBJS\s*=.*)$',
+                                '\\1 ../utilities/random.o',
+                                'Makefile')
+                # ../parcsr_ls/par_relax.o --> src/sstruct_ls/Makefile
+                with working_dir(join_path('src', 'sstruct_ls')):
+                    filter_file('^(OBJS\s*=.*)$',
+                                '\\1 ../parcsr_ls/par_relax.o',
+                                'Makefile')
+
+        # Set LDFLAGS - used for linking the shared library and for linking
+        # the tests.
+        make_args += ['LDFLAGS=%s' % ' '.join(ldflags)]
+
         # Hypre's source is staged under ./src so we'll have to manually
         # cd into it.
         with working_dir("src"):
             configure(*configure_args)
 
-            make()
+            make(*make_args)
             if self.run_tests:
-                make("check")
-                make("test")
+                make("check", *make_args)
+                make("test", *make_args)
                 Executable(join_path('test', 'ij'))()
                 sstruct = Executable(join_path('test', 'struct'))
                 sstruct()
                 sstruct('-in', 'test/sstruct.in.default', '-solver', '40',
                         '-rhsone')
-            make("install")
+            make("install", *make_args)
+
+    @property
+    def hypre_cuda_libs(self):
```

Comment:
For consistency, let’s do it through libs() when additional key “cuda” is provided. I think hdf5 uses suck keys to return different content

Reply:
Oh, yes, thanks for reminding me. I forgot we agreed to move this to the cuda package.

Reply:
I also forgot that ;-) i thought of keeping it in Hypre package, but moving to Cuda package makes much more sense